### PR TITLE
Feature - external custom tagger

### PR DESCRIPTION
**Pull request by tronxd**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -161,6 +163,11 @@ func tagCommand() *cli.Command {
 				Value:       cli.NewStringSlice(utils.GetAllTagGroupsNames()...),
 				DefaultText: "git,code2cloud",
 			},
+			&cli.StringFlag{
```
Comment (nimrodkor) [inline main.go:166]: @tronxd This is a workaround, it doesn't leverage the lib's [inner implementation of yaml loading](https://github.com/urfave/cli/blob/master/docs/v2/manual.md#values-from-alternate-input-sources-yaml-toml-and-others). Can we leverage that instead of writing all the yaml parsing ourselves?

File: src/common/runner/runner.go [hunk 2]
```diff
@@ -47,6 +49,9 @@ func (r *Runner) Init(commands *clioptions.TagOptions) error {
 		tagGroup.InitTagGroup(dir, commands.SkipTags)
 		if simpleTagGroup, ok := tagGroup.(*simple.TagGroup); ok {
 			simpleTagGroup.SetTags(extraTags)
+		} else if externalTagGroup, ok := tagGroup.(*external.TagGroup); ok && commands.ConfigFile != "" {
+			externalTagGroup.InitConfigFile(commands.ConfigFile)
+			externalTagGroup.InitExternalTagGroup()
```
Comment (nimrodkor) [inline src/common/runner/runner.go:54]: @tronxd Why split these methods? Can the runner only know 1 init method, and let the externalTagGroup manage this?

File: src/common/tagging/external/tag_group.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	config         map[interface{}]interface{}
+}
+
+func (t *TagGroup) InitConfigFile(configFilePath string) {
+	t.configFilePath = configFilePath
+}
+
+func (t *TagGroup) InitTagGroup(_ string, skippedTags []string) {
+	t.SkippedTags = skippedTags
+}
```
Comment (nimrodkor) [inline src/common/tagging/external/tag_group.go:25]: @tronxd We should load the tags from the conf file here, right?

**Review by nimrodkor — COMMENTED** [review]
@tronxd Nice progress!

File: src/common/runner/runner.go (hunk 2 above)
Comment (tronxd) [reply src/common/runner/runner.go:54]: @nimrodkor tagGroup.InitTagGroup only accepts the root dir and the skipped tags. do we want to add a variadic "data" argument to include (in our case) the config path?

File: src/common/tagging/external/tag_group.go (hunk 3 above)
Comment (tronxd) [reply src/common/tagging/external/tag_group.go:28]: Yes

File: main.go (hunk 1 above)
Comment (tronxd) [reply main.go:166]: @nimrodkor there is no parsing except what we want to parse internally (no YAML lib can parse more than just unmarshaling to a scheme

Comment (tronxd) [reply main.go:166]: @nimrodkor I will use this option only when binding it to the respected new CLI options

File: go.mod [hunk 4]
```diff
@@ -24,6 +24,7 @@ require (
 	go.opencensus.io v0.22.0
 	golang.org/x/sys v0.0.0-20210119212857-b64e53b001e4 // indirect
 	gopkg.in/validator.v2 v2.0.0-20200605151824-2b28d334fa05
+	gopkg.in/yaml.v2 v2.4.0
```
Comment (nimrodkor) [inline go.mod:27]: @tronxd Why do we need another yaml package, besides `github.com/sanathkr/yaml`? It does seem like this is the one we should use... Can we migrate over to `gopkg.in/yaml.v2`?

File: src/common/tagging/external/external_tag_group.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+tag_group:
```
Comment (nimrodkor) [inline src/common/tagging/external/external_tag_group.yml:1]: @tronxd Can this be in the tests / docs dir?

File: src/common/tagging/external/tag_group.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	filters      map[interface{}]interface{}
+	matches      []interface{}
+}
+
+func (t Tag) SatisfyFilters(block structure.IBlock, tagFilterDir string) bool {
+	newTags, existingTags := block.GetNewTags(), block.GetExistingTags()
+	newTags = append(newTags, existingTags...)
+	satisfyFilters := true
+	for filterKey, filterValue := range t.filters {
+		if filterKey == "tags" {
```
Comment (nimrodkor) [inline src/common/tagging/external/tag_group.go:37]: @tronxd Can be turned into a `switch` I believe

File: src/common/tagging/external/tag_group.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+type Tag struct {
+	tags.ITag
+	defaultValue string
+	filters      map[interface{}]interface{}
+	matches      []interface{}
+}
+
+func (t Tag) SatisfyFilters(block structure.IBlock, tagFilterDir string) bool {
```
Comment (nimrodkor) [inline src/common/tagging/external/tag_group.go:31]: @tronxd seems like a method we should look towards moving to the base tag package. not for this PR

File: src/common/tagging/external/tag_group.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+}
+
+func (t *TagGroup) InitTagGroup(dir string, skippedTags []string) {
+	t.SkippedTags = skippedTags
+	t.Dir = dir
+}
+
+func (t *TagGroup) InitExternalTagGroup() {
+	configMap := make(map[interface{}]interface{})
```
Comment (nimrodkor) [inline src/common/tagging/external/tag_group.go:76]: @tronxd Try to make this configMap a struct, with types. Would solve a lot of the unmarshalling you implemented yourself. Check out the example [here](https://github.com/go-yaml/yaml#example)

File: src/common/tagging/external/tag_group.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+						filteredNewTags = append(filteredNewTags[:i], filteredNewTags[i+1:]...)
+					}
+				}
+			} else {
+				filteredNewTags = append(filteredNewTags, tagValue)
+			}
+		}
+	}
+	filteredNewTags = append(filteredNewTags, existingTags...)
+	t.SetTags(filteredNewTags)
```
Comment (nimrodkor) [inline src/common/tagging/external/tag_group.go]: @tronxd Why is this necessary? This should be part of the setup, to support the skip functionality

File: src/common/clioptions/cli.go [hunk 10]
```diff
@@ -69,3 +72,18 @@ func validateOutput(v interface{}, _ string) error {
...
+	if v != "" {
+		val, ok := v.(string)
+		if !ok {
+			return validator.ErrUnsupported
+		}
+
+		if _, err := os.Stat(val); err != nil {
+			return fmt.Errorf("configuration file %s does not exist", v)
+		}
+
```
Comment (nimrodkor) [inline src/common/clioptions/cli.go:86]: @tronxd We should add more validations on this yaml, maybe not here 🤔

**Review by nimrodkor — COMMENTED** [review]
Very nicely done!

**Review by nimrodkor — APPROVED** [review]
Nice!
